doc/voq/lag_hld.md

```diff
@@ -0,0 +1,565 @@
+
+## 2.3 High Level Proposal
+The high level proposal is described here. The details are in the sections that follow. Each asic SONiC instance is responsible for LAG and LAG member ports that belong to that asic. Among other things this includes processing the configuration and keeping the application database updated with each LAG and its active member port list.
+*  The LAG configuration for an asic is limited to those LAGs which have its own network ports as members.
+*  Two new tables are defined and are a part of the Centralized Database CHASSIS_APP_DB. They are - "System LAG Table" and "System LAG Membership Table"
+*  The "System LAG Table" has an attribute called the "system_lag_id". This attribute will be automatically allocated (mechanism is described further down in this document).
+*  The SONiC network stack instance associated with an asic updates the entries in these tables for LAG and LAG Members that are controlled by that asic.
+*  All of the asic SONiC instances subscribe to these two tables and receive updates for remote LAG and LAG membership changes. These updates are programmed into the asic via the SAI APIs.
+Please note that the only per asic SONiC components that are aware of a remote LAG are SWSS and SYNCD. Other components (like teamd for example) are only aware of local LAG.
+
+## 3 Design Details
+
+### 3.1 Configuration
+
+* The **LAG configuration** in a Distributed VOQ system is same as is done in a non-VOQ system. No changes are required. LAG is expected to be configured on local ports only. No requirement to make system port specific LAG configuration. 
+* The name of the LAG in the PORTCHANNEL table is unique within the asic SONiC instance and must start with "PortChannel".
+* Also no changes are required for **LAG Member** configuration and **LAG Interface** configurations. PORTCHANNEL_MEMBER table and PORTCHANNEL_INTERFACE table are used in the same way how they are used in non-chassis system.
+
+### 3.2 Modules Design
+
+#### teamd: teamsyncd
+
+* teamsyncd only deals LAG on the port of its local asic instance. So no changes are required in teamsyncd. 
+
+#### orchagent: portsorch
+**Local LAG**: portsorch obtains a system wide unique system_lag_id from centralized database CHASSIS_APP_DB. portsorch sends this system_lag_id in SAI_LAG_ATTR_SYSTEM_PORT_AGGREGATE_ID attribute to SAI while creating LAG in SAI. After creating local LAG, the PortChannel with switch_id and system_lag_id is written to the SYSTEM_LAG_TABLE in centralized database CHASSIS_APP_DB. A unique name is assigned automatically to each LAG in the system lag table in the CHASSIS_APP_DB using the following naming scheme (similar to the nameing scheme used for naming the system ports).
+```
+<Host name>|<Asic name>|<port channel name in its local asic instance>
+```
+The _Host name_ and _Asic name_ are available in CONFIG_DB device meta data attributes **hostnmae** and **asic_name** respectively.
```

Comment:
`asic_name` is only available on multi-asic devices. On single-asic devices the asic can be identified by the `hostname` only.

Do you plan to make `asic_name` available on single-asic devices as well? Or can a lag name simply omit <Asic name> for single-asic devices, i.e. `Slot1|PortChannel001`?

Reply:
As per current system lag implementation, if the switch type is "voq", "asic_name" is mandatory in DEVICE_METADATA['localhost'] and it should have  a valid non-zero sized value. So even for the single asic linecards, "asic_name" must be provided with a valid value.

Reply:
I see that it became mandatory when #1605 merged (https://github.com/Azure/sonic-swss/pull/1605/files#diff-3cba33a216c0ea2b942dc03b2638fa27a40a8b6eb81abf9f83f13e1ee7831ac4R231-R237), i.e.:
```cpp
    if (!cfgDeviceMetaDataTable.hget("localhost", "asic_name", value))
    {
        // asic_name is not configured.
        SWSS_LOG_ERROR("Asic name is not configured");
        return false;
    }
```

However I'm a bit surprised as this basically prevents system ports to be configured without `asic_name`. This doesn't seem necessary.

What is the reason to make `asic_name` mandatory even on single-asic devices?

Reply:
I agree. "asic_name" for single asic seems unnecessary. 
However, there is a reason why we need asic_name as mandatory in the system lag name. We use "|" as separator between the hostname, asic_name and port name. In the Distributed VOQ LAG HLD design discussion meeting it was told that in  the future all tables will have "|" as the table separator. So for the system lag member table the key will look something like 
**"Linecard1|Asic0|PortChannel001|Linecard1|Asic0|Ethernet1"**. This is **\<derived system lag name\>|\<system port name\>**. In the lag member processing, to extract the port channel alias and member system port name alias from the key, it was suggested to use the fixed format of system lag and system port names. So we do not want two different formats.
Please note that for now, the separator is ":" and porchannel alias and system port member alias extraction is based on this. But in the future this may not be the case.

Reply:
Ok. Then I'll make sure that `asic_name` is set for VoQ regardless of of the number of asics in sonic-buildimage [PR5997](https://github.com/Azure/sonic-buildimage/pull/5991).

Reply:
I assume this is for single-chip linecards that are in a chassis with other chips. I don't think it's unreasonable to expect table entries are qualified with 'slotname|chipname' always. It's a little clunky for single-chip linecards, but not terrible and if it helps simplify code, we can do that. 

For single-chip fixed systems that are not a in chassis, we shouldn't set the switch_type to 'voq' anyway.